### PR TITLE
[WIP] Assume environments are channels-first by default

### PR DESCRIPTION
**Pull request by qxcv**
## Description
This change removes all automatic image transposes, including automatic application of `VecTranspose`. Instead, environments are assumed to be channels-first; users with channels-last environments can wrap them with `VecTranspose` before passing them to SB3 algorithms. It also makes a few other changes that are necessary for compatibility with channels-first environments:

- `VecFrameStack` is updated to support channels-first environments (or channels in any position, really).
- Atari environment constructors now automatically wrap with `VecTranspose`.

## Motivation and Context

This PR is one way of addressing the feature request in #142. It also resolves the original confusion behind #103.

- [x] I have raised an issue to propose this change ([required](https://github.com/DLR-RM/stable-baselines3/blob/master/CONTRIBUTING.md) for new features and bug fixes)

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation (update in the documentation)

(this is not really a bug fix or a new feature, but rather a change of functionality)

## Checklist:

- [x] I've read the [CONTRIBUTION](https://github.com/DLR-RM/stable-baselines3/blob/master/CONTRIBUTING.md) guide (**required**)
- [x] I have updated the changelog accordingly (**required**).
- [x] My change requires a change to the documentation.
- [x] I have updated the tests accordingly (*required for a bug fix or a new feature*).
- [x] I have updated the documentation accordingly.
- [x] I have reformatted the code using `make format` (**required**)
- [x] I have checked the codestyle using `make check-codestyle` and `make lint` (**required**)
- [x] I have ensured `make pytest` and `make type` both pass. (**required**)

### CODE REVIEW

File: docs/conf.py [hunk 1]
```diff
@@ -52,7 +52,7 @@ def __getattr__(cls, name):
 # -- Project information -----------------------------------------------------
 
 project = "Stable Baselines3"
-copyright = "2020, Stable Baselines3"
+copyright = "2020, Stable Baselines3"  # noqa: A001
```
Comment (Miffyli) [inline docs/conf.py:55]: Not sure if we want this kind of edits. @araffin ?

File: stable_baselines3/common/vec_env/vec_frame_stack.py [hunk 2]
```diff
@@ -13,15 +13,22 @@ class VecFrameStack(VecEnvWrapper):
 
     :param venv: the vectorized environment to wrap
     :param n_stack: Number of frames to stack
+    :param dim: Dimension along which to stack frames (not including batch dimension).
```
Comment (Miffyli) [inline stable_baselines3/common/vec_env/vec_frame_stack.py:16]: Could mention here it also supports the negative-dimensions as slicing/indexing usually does.

File: stable_baselines3/common/vec_env/vec_transpose.py [hunk 3]
```diff
@@ -46,6 +46,13 @@ def transpose_image(image: np.ndarray) -> np.ndarray:
 
     def step_wait(self) -> VecEnvStepReturn:
         observations, rewards, dones, infos = self.venv.step_wait()
+        terminal_key = "terminal_observation"
```
Comment (Miffyli) [inline stable_baselines3/common/vec_env/vec_transpose.py:49]: Good catch. This seems to have been a bug in original version? If so, please mention in the docs that it fixes this.

File: stable_baselines3/common/base_class.py [hunk 4]
```diff
@@ -167,10 +166,6 @@ def _wrap_env(self, env: GymEnv) -> VecEnv:
                 print("Wrapping the env in a DummyVecEnv.")
             env = DummyVecEnv([lambda: env])
 
-        if is_image_space(env.observation_space) and not isinstance(env, VecTransposeImage):
```
Comment (Miffyli) [inline stable_baselines3/common/base_class.py:170]: This check should be kept but with an exception or a warning if last dimension seems to be channels (e.g. `assert shape[0] < shape[2]`. The whole "channels-first" convention will throw some people off, and in worst cases nothing else will warn about this, apart from CNN layers going to negative dimension shapes :)

File: stable_baselines3/common/policies.py [hunk 5]
```diff
@@ -229,20 +228,6 @@ def predict(
         #     mask = [False for _ in range(self.n_envs)]
         observation = np.array(observation)
 
-        # Handle the different cases for images
-        # as PyTorch use channel first format
-        if is_image_space(self.observation_space):
```
Comment (Miffyli) [inline stable_baselines3/common/policies.py:234]: Same check as with base_class (see if channels are first, and throw warnings/exceptions if not).

**Comment by Miffyli** [comment]
Sorry for the delay. Overall looks good, left some nitpicky comments and I will still go over the docs etc to see that nothing about old behaviour remains :)

File: docs/guide/examples.rst [hunk 6]
```diff
@@ -287,6 +287,8 @@ and multiprocessing for you.
       obs, rewards, dones, info = env.step(action)
       env.render()
 
+Note that ``make_atari_env`` implicitly applies a ``VecTransposeImage`` wrapper that puts channels along the leading axis, rather than the (default) final axis. If you want to use Stable Baselines 3 with a different image-based environment that puts channels last, then you will have to wrap the environment with a ``VecTransposeImage`` before passing the environment to any Stable Baselines 3 algorithms.
```
Comment (Miffyli) [inline docs/guide/examples.rst:290]: Hmm I do not see where VecTransposeImage is being applied on Atari envs. Still to-be-added?

**Comment by araffin** [comment]
As mentioned in the issue, I would rather add a check to know if the environment is channel-first or channel-last (`shape[0] < shape[2]` should work in most cases) and still wrap it automatically if needed (otherwise, this would break code).
This would still require changes in `VecFrameStack` and additional checks in `common/policies.py` (of rthe `predict()` method).

**Comment by Miffyli** [comment]
Bit of time since the last update on this one. @qxcv Do you mind if I hijack this enhancement and develop it (perhaps make a new branch for it too if easier)? 

@araffin I agree that solution could be the best for now (only wrap if needed but still automatically do it if necessary, and update necessary parts).

**Comment by qxcv** [comment]
Sure, that's fine. Probably easier to make a new branch.

**Comment by Miffyli** [comment]
@qxcv I made it in #213 where I also thank you in the changelog (I assume that is okey to you :) ). I will leave this PR and related issue open though as simpler code without transposing would be the most ideal, if in any way possible.

**Pull request closed without merging**